File: apps/bridge/test/defaultWorkspace.test.ts
/**
 * Default workspace resolver: repo in a checkout, Documents/Grok otherwise.
 */

import assert from "node:assert/strict";
import { mkdirSync, mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { describe, it } from "node:test";
import {
  PRODUCT_WORKSPACE_FOLDER,
  REPO_MARKER,
  findRepoRoot,
  productionWorkspaceDir,
  resolveDefaultWorkspaceCwd,
  userDocumentsDir,
} from "../src/defaultWorkspace.js";

/**
 * Build a fake monorepo with the same marker the product walk looks for.
 * @returns Absolute fake repo root.
 */
function fakeRepo(): string {
  const root = mkdtempSync(path.join(tmpdir(), "grok-repo-"));
  const markerDir = path.join(root, path.dirname(REPO_MARKER));
  mkdirSync(markerDir, { recursive: true });
  writeFileSync(path.join(root, REPO_MARKER), "// marker\n");
  return root;
}

describe("findRepoRoot", () => {
  it("walks up from a nested path to the marker", () => {
    const root = fakeRepo();
    const nested = path.join(root, "apps", "desktop", "src");
    mkdirSync(nested, { recursive: true });
    assert.equal(findRepoRoot(nested), root);
    assert.equal(findRepoRoot(root), root);
  });

  it("returns null when no checkout is nearby", () => {
    const bare = mkdtempSync(path.join(tmpdir(), "grok-bare-"));
    assert.equal(findRepoRoot(bare), null);
  });
});

describe("userDocumentsDir / productionWorkspaceDir", () => {
  it("uses HOME/Documents/Grok on every OS", () => {
    const home = "/Users/dev";
    const env = { HOME: home, USERPROFILE: home };
    assert.equal(userDocumentsDir(env), path.join(home, "Documents"));
    assert.equal(
      productionWorkspaceDir(env),
      path.join(home, "Documents", PRODUCT_WORKSPACE_FOLDER),
    );
  });

  it("prefers XDG_DOCUMENTS_DIR when set", () => {
    const env = {
      HOME: "/home/dev",
      XDG_DOCUMENTS_DIR: "/mnt/docs",
    };
    assert.equal(userDocumentsDir(env), path.resolve("/mnt/docs"));
    assert.equal(
      productionWorkspaceDir(env),
      path.join(path.resolve("/mnt/docs"), PRODUCT_WORKSPACE_FOLDER),
    );
  });

  it("falls back to USERPROFILE when HOME is empty", () => {
    const env = { USERPROFILE: "C:\\Users\\dev" };
    assert.equal(
      userDocumentsDir(env),
      path.join("C:\\Users\\dev", "Documents"),
    );
  });
});

describe("resolveDefaultWorkspaceCwd", () => {
  it("BRIDGE_CWD wins over repo and Documents", () => {
    const root = fakeRepo();
    const cwd = resolveDefaultWorkspaceCwd({
      startDir: root,
      env: {
        BRIDGE_CWD: "/tmp/explicit-ws",
        HOME: "/Users/dev",
      },
    });
    assert.equal(cwd, path.resolve("/tmp/explicit-ws"));
  });

  it("uses the monorepo root when running from a checkout", () => {
    const root = fakeRepo();
    const cwd = resolveDefaultWorkspaceCwd({
      startDir: path.join(root, "apps", "bridge", "src"),
      env: { HOME: "/Users/dev" },
    });
    assert.equal(cwd, root);
  });

  it("uses Documents/Grok when no checkout is nearby", () => {
    const bare = mkdtempSync(path.join(tmpdir(), "grok-prod-"));
    const home = mkdtempSync(path.join(tmpdir(), "grok-home-"));
    const cwd = resolveDefaultWorkspaceCwd({
      startDir: bare,
      env: { HOME: home, USERPROFILE: home },
    });
    assert.equal(
      cwd,
      path.join(home, "Documents", PRODUCT_WORKSPACE_FOLDER),
    );
  });
});

File: apps/shell/default_workspace_test.go
package main

import (
	"os"
	"path/filepath"
	"testing"
)

func TestUserDocumentsDirUnifiedLeaf(t *testing.T) {
	t.Setenv("XDG_DOCUMENTS_DIR", "")
	home := filepath.Join(t.TempDir(), "home")
	got := UserDocumentsDir(home)
	want := filepath.Join(home, "Documents")
	if got != want {
		t.Fatalf("got %q want %q", got, want)
	}
	prod := ProductionWorkspaceDir(home)
	if prod != filepath.Join(want, ProductWorkspaceFolder) {
		t.Fatalf("prod %q", prod)
	}
}

func TestUserDocumentsDirPrefersXDG(t *testing.T) {
	docs := t.TempDir()
	t.Setenv("XDG_DOCUMENTS_DIR", docs)
	if got := UserDocumentsDir("/ignored"); got != docs {
		t.Fatalf("got %q want %q", got, docs)
	}
}

func TestDefaultBridgeCWDUsesRepoRoot(t *testing.T) {
	repo := filepath.Join(t.TempDir(), "grok-desktop")
	if got := DefaultBridgeCWD(repo); got != repo {
		t.Fatalf("got %q want repo %q", got, repo)
	}
}

func TestDefaultBridgeCWDUsesDocumentsWhenNoRepo(t *testing.T) {
	home := t.TempDir()
	t.Setenv("XDG_DOCUMENTS_DIR", "")
	t.Setenv("HOME", home)
	t.Setenv("USERPROFILE", home)
	got := DefaultBridgeCWD("")
	want := filepath.Join(home, "Documents", ProductWorkspaceFolder)
	if got != want {
		t.Fatalf("got %q want %q", got, want)
	}
	if st, err := os.Stat(got); err != nil || !st.IsDir() {
		t.Fatalf("production workspace should be created: err=%v", err)
	}
}
